### PR TITLE
Update stress calculation for MACE

### PR DESCRIPTION
**Pull request by sauradeep93**
MACE requires the stress to be computed directly rather than through the 'stress volume'

### CODE REVIEW

**Comment by ajhoffman1229** [comment]
@swagataroy123 can you add your comment/review here instead of the private one?

File: nff/io/ase_calcs.py [hunk 1]
```diff
@@ -197,15 +210,27 @@ def calculate(
...
+                
+                self.results["stress"] = full_3x3_to_voigt_6_stress(
+                    torch.mean(prediction["stress"], dim=0).cpu().numpy()*self.energy_units_to_eV/1.0)#1.0 as volume is in A**3, but conversion for energy is needed
+            else:  #for other models
+                stress = prediction["stress_volume"].detach().cpu().numpy() * (1 / const.EV_TO_KCAL_MOL) # TODO change to more general prediction
+                self.results["stress"] = stress * (1 / atoms.get_volume())
             if "stress_disp" in prediction:
                 self.results["stress"] = self.results["stress"] + prediction["stress_disp"]
-            self.results["stress"] = full_3x3_to_voigt_6_stress(self.results["stress"])
+            #self.results["stress"] = full_3x3_to_voigt_6_stress(self.results["stress"])#for mace model, this line is already implemented above, for other models uncomment this line
```
Comment (swagataroy123) [inline nff/io/ase_calcs.py:231]: Hi, I have a suggestion. Instead of commenting/uncommenting it, we can have
self.results["stress"] = torch.mean(prediction["stress"], dim=0).cpu().numpy()*self.energy_units_to_eV/1.0
and continue transforming in the next stage like other models

Comment (sauradeep93) [reply nff/io/ase_calcs.py:231]: Makes sense, so basically you are suggesting to remove lines 224 and 225, and instead keep 231 as it is (in its uncommented form)

File: nff/io/ase_calcs.py [hunk 2]
```diff
@@ -85,6 +86,18 @@ def __init__(
...
+        
+        # Unit conversion factors
+        self.energy_units_to_eV = 1.0  # assuming default model outputs energy in eV
+        self.length_units_to_A = 1.0   # assuming default model uses A**3 for length
+
+        # Adjust conversion factors if the model uses different units
+        if self.model_units == "kcal/mol":
+            self.energy_units_to_eV = 0.0433641  # kcal/mol -> eV
+        if self.prediction_units == "nm":
+            self.length_units_to_A = 10.0  # nm -> AA*3
```
Comment (ajhoffman1229) [inline nff/io/ase_calcs.py:99]: It might be best to define these unit conversions near the beginning of the file, similar to the constant `HARTREE_TO_EV`, although some of these conversions might already be defined in the ASE `units` module or the NFF `constants` module, which could then be imported and used for clarity.

File: nff/io/ase_calcs.py (hunk 1 above)
Comment (swagataroy123) [reply nff/io/ase_calcs.py:231]: Yes but line 225 stays without the transformation as shown in previous comment

File: nff/io/ase_calcs.py [hunk 3]
```diff
@@ -197,15 +210,27 @@ def calculate(
...
+                    try:
+                        prediction["stress"] = prediction["stress"].reshape(3, 3)
+                       # print("Reshaped stress to 3*3:", prediction["stress"])
+                    except ValueError as e:
+                        raise ValueError(f"Error reshaping stress tensor: {e}, shape: {prediction['stress'].shape}")
+
+               # print("The current dimensions of the stress tensor is:", prediction["stress"].ndim)
+                
+                self.results["stress"] = full_3x3_to_voigt_6_stress(
+                    torch.mean(prediction["stress"], dim=0).cpu().numpy()*self.energy_units_to_eV/1.0)#1.0 as volume is in A**3, but conversion for energy is needed
```
Comment (ajhoffman1229) [inline nff/io/ase_calcs.py:225]: I am not sure I understand why you are dividing by 1.0? You can comment that you are converting to eV/&angst;<sup>3</sup> and I don't think further explanation is needed beyond that and the 1.0 is redundant.

File: nff/io/ase_calcs.py [hunk 4]
```diff
@@ -197,15 +210,27 @@ def calculate(
             self.results["embedding"] = embedding
 
         if requires_stress:
-            stress = prediction["stress_volume"].detach().cpu().numpy() * (
-                1 / const.EV_TO_KCAL_MOL
-                # TODO change to more general prediction
-            )
-            self.results["stress"] = stress * (1 / atoms.get_volume())
+            if isinstance(self.model, NffScaleMACE):#the implementation of stress calculation in MACE is a bit different and hence this is needed (ASE_suit: mace/mace/calculators/mace.py)
+               # print(f"Original stress shape: {prediction['stress'].shape}")
```
Comment (ajhoffman1229) [inline nff/io/ase_calcs.py:214]: Please remove commented print statements like these before merging

File: nff/io/ase_calcs.py (hunk 1 above)
Comment (ajhoffman1229) [reply nff/io/ase_calcs.py:231]: Yeah it is never a good idea to have a line that someone needs to uncomment for the code to work if they're using a different model. You also don't want to remove 224-225 entirely, just the conversion to the Voigt stress format (with `full_3x3_to_voigt_6_stress`). The code should instead read:

```python
            if isinstance(self.model, NffScaleMACE):
                if prediction["stress"].ndim==1:
                    try:
                        prediction["stress"] = prediction["stress"].reshape(3, 3)
                       # print("Reshaped stress to 3*3:", prediction["stress"])
                    except ValueError as e:
                        raise ValueError(f"Error reshaping stress tensor: {e}, shape: {prediction['stress'].shape}")

                self.results["stress"] = torch.mean(prediction["stress"], dim=0).cpu().numpy()*self.energy_units_to_eV) # convert to eV/Angstrom^3
           else:  #for other models
                stress = prediction["stress_volume"].detach().cpu().numpy() * (1 / const.EV_TO_KCAL_MOL) # TODO change to more general prediction
                self.results["stress"] = stress * (1 / atoms.get_volume())
            if "stress_disp" in prediction:
                self.results["stress"] = self.results["stress"] + prediction["stress_disp"]
            self.results["stress"] = full_3x3_to_voigt_6_stress(self.results["stress"])
            atoms.results = self.results.copy()
```

**Review by ajhoffman1229 — CHANGES_REQUESTED** [review]
I think @swagataroy123's suggestion is correct and needs to be made before this PR is merged. I also had a few stylistic changes to suggest to make the code more readable.

**Review by swagataroy123 — COMMENTED** [review]
`conversion_factor: dict = const.conversion_factors.get((self.model_units, self.prediction_units), const.DEFAULT) `Line 174
Please make sure that this function does not already converts stress prediction for you. If not please incorporate conversion of stress here to maintain continuity of code instead of
Line 217:                    ` torch.mean(prediction["stress"], dim=0).cpu().numpy()*self.energy_units_to_eV #converting to eV/Angstrom^3`
Then you will not need to use
Line 90 - 94

**Review by swagataroy123 — COMMENTED** [review]
I checked the constant.convert_units code and it does not take care of stress. @ajhoffman1229 , do you use some other codes in htvs for stress conversion units for VASP?
If not we can modify the dictionaries in the constants.py file to incorporate stress as well for conversion. 
It can be through this pull request as it will also change this implementation
Same with https://github.com/learningmatter-mit/NeuralForceField/blob/aa4d509dff8d7af31716dbc3f30d70c83597697e/nff/io/ase_calcs.py#L200
These conversions can be done in one go.

**Comment by ajhoffman1229** [comment]
@swagataroy123 
> I checked the constant.convert_units code and it does not take care of stress. @ajhoffman1229 , do you use some other codes in htvs for stress conversion units for VASP?
Yes, there is some code for the conversion in htvs. I will send a link in Slack (since it's a private repo).

File: nff/io/ase_calcs.py [hunk 5]
```diff
@@ -283,10 +299,8 @@ def offset_energy(self, atoms, energy: Union[float, np.ndarray]):
         for ele, num in ads_count.items():
             ref_en += num * stoidict.get(ele, 0.0)
         ref_en += stoidict.get("offset", 0.0)
-        if self.offset_units == "atomic":
```
Comment (swagataroy123) [inline nff/io/ase_calcs.py:286]: @sauradeep93 maybe some unintentional change?

File: nff/io/ase_calcs.py [hunk 6]
```diff
@@ -454,13 +468,11 @@ def set(self, **kwargs):
 
         The special keyword 'parameters' can be used to read
         parameters from a file."""
-        changed_params = Calculator.set(self, **kwargs)
```
Comment (swagataroy123) [inline nff/io/ase_calcs.py:457]: same here? @sauradeep93 this maybe ruff or something

Comment (swagataroy123) [reply nff/io/ase_calcs.py:457]: ok so I checked this two changes. they are from @xiaochendu. He can comment better

Comment (xiaochendu) [reply nff/io/ase_calcs.py:457]: Oh boy, I have these deleted lines in my code. Can we revert these?

File: nff/io/ase_calcs.py (hunk 5 above)
Comment (xiaochendu) [reply nff/io/ase_calcs.py:286]: Same here, maybe these changes weren't reflected in @sauradeep93's local repo?

File: nff/io/ase_calcs.py (hunk 6 above)
Comment (swagataroy123) [reply nff/io/ase_calcs.py:457]: can u suggest how i think ruff did it for him or something

Comment (xiaochendu) [reply nff/io/ase_calcs.py:457]: 😂 I think the easiest way might be to copy these lines back? But strange, I'm not sure why ruff would do something like this.

Comment (xiaochendu) [reply nff/io/ase_calcs.py:457]: Also @ajhoffman1229 might have better ideas

File: nff/io/ase_calcs.py (hunk 5 above)
Comment (swagataroy123) [reply nff/io/ase_calcs.py:286]: oh ok

File: nff/io/ase_calcs.py (hunk 6 above)
Comment (swagataroy123) [reply nff/io/ase_calcs.py:457]: i guess then it's what u thought. He didn't have it or something. let's see lol

Comment (sauradeep93) [reply nff/io/ase_calcs.py]: No idea how these changes came across, because I didn't do them. There could be a difference between my local repo and the GitHub repo. Anyways I copy pasted the stuff as it was before. Please have a look and let me know incase I missed something.
Sorry for the inconvenience

File: nff/utils/constants.py [hunk 7]
```diff
@@ -61,6 +61,57 @@
     "_grad": 1.0,
 }
 
+
```
Comment (xiaochendu) [inline nff/utils/constants.py:64]: Seems like the following are repeats of the above but with additional keys. Can we merge them, i.e., delete the ones on top?

File: nff/io/ase_calcs.py [hunk 8]
```diff
@@ -283,8 +291,9 @@ def offset_energy(self, atoms, energy: Union[float, np.ndarray]):
         for ele, num in ads_count.items():
             ref_en += num * stoidict.get(ele, 0.0)
         ref_en += stoidict.get("offset", 0.0)
+
         if self.offset_units == "atomic":
-            energy += ref_en * HARTREE_TO_EV
+             energy += ref_en * HARTREE_TO_EV
```
Comment (xiaochendu) [inline nff/io/ase_calcs.py:296]: There seems to be a small indentation mismatch here? Does the code run without errors in this version?

File: nff/io/ase_calcs.py (hunk 6 above)
Comment (xiaochendu) [reply nff/io/ase_calcs.py]: Thanks! I just added 2 additional (small) comments!

File: nff/utils/constants.py (hunk 7 above)
Comment (sauradeep93) [reply nff/utils/constants.py:34]: Done !!

**Review by xiaochendu — APPROVED** [review]
Thank you for making the changes!

**Review by swagataroy123 — APPROVED** [review]

**Review by ajhoffman1229 — APPROVED** [review]
